File: src/mobile/ios/SaaSPortal/ViewModels/DashboardViewModel.swift
import Foundation

/// ダッシュボード画面の UI 状態。
enum DashboardUiState {
    case loading
    case success(DashboardSummary, UsageTrendResponse?)
    case error(String)
}

/// ダッシュボード画面の状態管理 ViewModel。
@Observable
final class DashboardViewModel {
    private(set) var uiState: DashboardUiState = .loading
    private let portalService: PortalServiceProtocol

    init(portalService: PortalServiceProtocol = PortalService()) {
        self.portalService = portalService
    }

    /// ダッシュボードサマリーを取得し、成功後にトレンドも取得する。
    @MainActor
    func loadDashboard() async {
        uiState = .loading
        do {
            let summary = try await portalService.getDashboardSummary()
            uiState = .success(summary, nil)
            // トレンドは非同期で追加読み込み
            await loadTrends(summary: summary)
        } catch {
            uiState = .error(errorMessage(from: error))
        }
    }

    /// 月次利用推移データを取得する。
    @MainActor
    func loadTrends(summary: DashboardSummary? = nil) async {
        let currentSummary: DashboardSummary
        if let summary {
            currentSummary = summary
        } else if case .success(let existing, _) = uiState {
            currentSummary = existing
        } else {
            return
        }

        do {
            let trends = try await portalService.getUsageTrends()
            uiState = .success(currentSummary, trends)
        } catch {
            // トレンド取得失敗はサマリーを維持する
            uiState = .success(currentSummary, nil)
        }
    }

    // MARK: - Private

    private func errorMessage(from error: Error) -> String {
        if let apiError = error as? APIError {
            return apiError.localizedDescription
        }
        return error.localizedDescription
    }
}

File: src/mobile/android/app/src/main/kotlin/com/example/saasportal/ui/dashboard/DashboardViewModel.kt
package com.example.saasportal.ui.dashboard

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.saasportal.domain.model.DashboardSummary
import com.example.saasportal.domain.model.UsageTrend
import com.example.saasportal.domain.repository.PortalRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class DashboardViewModel @Inject constructor(
    private val portalRepository: PortalRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow<DashboardUiState>(DashboardUiState.Loading)
    val uiState: StateFlow<DashboardUiState> = _uiState.asStateFlow()

    init {
        loadDashboard()
    }

    fun loadDashboard() {
        viewModelScope.launch {
            _uiState.value = DashboardUiState.Loading
            try {
                val summaryDeferred = async { portalRepository.getDashboardSummary() }
                val trendsDeferred = async { portalRepository.getUsageTrends() }
                val summary = summaryDeferred.await()
                val trends = trendsDeferred.await()
                _uiState.value = DashboardUiState.Success(
                    summary = summary,
                    trends = trends,
                )
            } catch (e: Exception) {
                _uiState.value = DashboardUiState.Error(
                    message = e.message ?: "データの取得に失敗しました",
                )
            }
        }
    }
}

sealed interface DashboardUiState {
    data object Loading : DashboardUiState
    data class Success(
        val summary: DashboardSummary,
        val trends: List<UsageTrend>,
    ) : DashboardUiState
    data class Error(val message: String) : DashboardUiState
}
